go-controller/pkg/allocator/pod/pod_annotation.go

```diff
@@ -59,9 +106,51 @@ func (allocator *PodAnnotationAllocator) AllocatePodAnnotation(
 		allocator.netInfo,
 		pod,
 		network,
+		nil,
 		reallocateIP,
 	)
+}
+
+func (allocator *PersistentIPsPodAnnotationAllocator) AllocatePodAnnotation(
+	ipAllocator subnet.NamedAllocator,
+	pod *v1.Pod,
+	network *nadapi.NetworkSelectionElement,
+	reallocateIP bool) (
+	*v1.Pod,
+	*util.PodAnnotation,
+	error) {
 
+	ipamClaim, err := allocator.ipamClaimsFetcher.FindIPAMClaim(network)
```

Comment:
The client is supposed to be using the VM/pod's name when creating the IPAM as part of the key. 

Thus, I think this scenario you pose is not something that would happen. 

We are currently only rolling back if the pod's annotation is not updated - the motivation is if we fail to update the IPAMClaim's status with its assigned IPs, we return an error, and in the next reconcile cycle we carry on from the IPs already available in the annotation, and then try again to persist the IPs in the claim. 

If we want to protect against the scenario you pose, I guess we should rollback, but I'm afraid the "second" pod would keep crash-looping until deleted (since it will be trying to update the claim with IPs that are different from the ones it already has).

Reply:
> The client is supposed to be using the VM/pod's name when creating the IPAM as part of the key.
> 
> Thus, I think this scenario you pose is not something that would happen.

I don't understand your reasoning on why this would not happen.

> We are currently only rolling back if the pod's annotation is not updated - the motivation is if we fail to update the IPAMClaim's status with its assigned IPs, we return an error, and in the next reconcile cycle we carry on from the IPs already available in the annotation, and then try again to persist the IPs in the claim.

Yes, but in the meantime, another pod pointing to the same ipam claim reference can have a different set of IPs allocated.

> If we want to protect against the scenario you pose, I guess we should rollback, but I'm afraid the "second" pod would keep crash-looping until deleted (since it will be trying to update the claim with IPs that are different from the ones it already has).

The second pod would use the IPs that are already set on the IPAM claim and also I did not see any code preventing the update.

Reply:
> > The client is supposed to be using the VM/pod's name when creating the IPAM as part of the key.
> > Thus, I think this scenario you pose is not something that would happen.
> 
> I don't understand your reasoning on why this would not happen.

Let me try to explain this with an example. Let's refer to a kubevirt VM named vm1.

KubeVirt will schedule a pod named `vm1-launcher-xyz` for it. It will also create an IPAMClaim annotation named ``vm1-launcher-xyz.tenantblue` (I'm assuming we only have a single network in the vm, named tenantblue).

The point is the IPAMClaims are not to be shared between pods.

> 
> > We are currently only rolling back if the pod's annotation is not updated - the motivation is if we fail to update the IPAMClaim's status with its assigned IPs, we return an error, and in the next reconcile cycle we carry on from the IPs already available in the annotation, and then try again to persist the IPs in the claim.
> 
> Yes, but in the meantime, another pod pointing to the same ipam claim reference can have a different set of IPs allocated.
> 

Right but, again, the IPAMClaims are not meant to be shared between pods. I guess I would reach a similar stateif I created a pod with the OVN annotations of some other running (or about to be running) pod.

> > If we want to protect against the scenario you pose, I guess we should rollback, but I'm afraid the "second" pod would keep crash-looping until deleted (since it will be trying to update the claim with IPs that are different from the ones it already has).
> 
> The second pod would use the IPs that are already set on the IPAM claim and also I did not see any code preventing the update with an error.

Right, when you wrote this that particular code didn't exist.

Reply:
Please confirm one thing for me;

> Also, shouldn't we rollback the ip allocation if updating the ipam claim fails because of a conflict?

If we rollback the IP allocation when updating the IPAMClaim fails because of a conflict - but the pod annotation was successfully updated - and we throw an error, the pod's annotation would be left updated.

It would error, and in a subsequent reconcile cycle we would carry-over the IPs from the annotation and try to use them again (thus allocating the IP on the IPAllocator, and attempting again to update the claims).

Am I missing something ?

Reply:
It looks straight forward if you move the ipam claim stuff to the `allocatePorAnnotationWithRollback` function and you actually may have less code duplication.

Reply:
Pushed in a new version with this.

I have mixed feelings, since I've had to change some tests a little bit, but overall it is less code, and probably more scoped. 

Let me know what you think. 

Question: the rollback should happen **only** in a conflict ? (to prevent the scenario you pose)